### PR TITLE
Add CLI user commands

### LINKED ISSUE #2136: Add CLI user commands
### Is your feature request related to a specific problem?

There is currently no way to authenticate via the CLI or do anything else related to users/user management

### Describe the solution you'd like

Some kind of `fides login` that would enable the user to login and store those credentials locally for continued use

### Additional context

This ticket must be completed before we can lock down all of the API endpoints

Mostly looking at the [solution used by the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)

### PR DESCRIPTION
**Pull request by ThomasLaPiana**
Closes #2136 

### Code Changes

* [x] add a `fides user login` CLI command (uses `user` endpoint, not `client` endpoint)
  * [x] add logic for getting an access_token from the webserver with a username/password or with an existing credentials file
  * [x] add logic for storing the access_token and the username in a `.fides_credentials` file in the user directory
  * [x] Base64 encode the password before sending it
* [x] add a `fides user create` command for creating a user
* [x] add a `fides user permissions` command for showing the scopes of the current user
* [x] add tests for all of the above
* [x] use matching username/password options for commands that require it (`login` and `create`, for example)
* [x] allow for credentials file path override via env var

### Relevant Links:

* [oauth docs](https://docs.ethyca.com/fides/dsr_quickstart/dsr_support/oauth)

### Steps to Confirm

* [ ] make sure the webserver is available when running your commands
* [ ] run `fides db reset -y` to clear the database of any existing users
* [ ] run `fides user login -u root_user -p Testpassword1!` and verify a credentials file is written with the correct info
* [ ] run `fides user login` and verify that you're prompted for the username and password, and that the password is hidden
* [ ] run `fides user login -u fidesadmin -p wrongpassword` and verify the error message
* [ ] run `fides user create newuser HolyCow1*1` and confirm that a new user is created
* [ ] run `fides user login -u newuser -p HolyCow1*1` and confirm that the login is successful and the credentials file is updated
* [ ] run `fides user permissions` and verify that the permissions are correct (all permissions)
* [ ] set the `FIDES_CREDENTIALS_PATH` env var and confirm that is respected as the path for the credentials file

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [ ] Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created (#2165 )
* [x] Update `CHANGELOG.md`

### Description Of Changes

This PR adds a `user` command group to the CLI that allows the CLI to authenticate with the webserver and store the `username`, `password`, `access_token` and `id` in a local `.fides_credentials` file

### CODE REVIEW

**Comment by ThomasLaPiana** [comment]
Bug Found: root client can't get their own scopes

**Comment by ThomasLaPiana** [comment]
Potential bug found: I created a new user but when I search for that user, the search turns up empty

**Comment by ThomasLaPiana** [comment]
@sanders41 do you think I should be using clients or users here? I'm thinking users is what we actually want, since clients are ephemeral API users _created_ by actual users right?

**Comment by sanders41** [comment]
> @sanders41 do you think I should be using clients or users here? I'm thinking users is what we actually want, since clients are ephemeral API users created by actual users right?

Right, I would use users

**Comment by ThomasLaPiana** [comment]
running into a perplexing error:

```
fides> fides user create newusera Newpass!word1
Loading config from: .fides/fides.toml
{'username': 'newusera', 'password': 'Newpass!word1'}
{ 'detail': [ { 'loc': ['body', 'password'],
                'msg': "'utf-8' codec can't decode byte 0xec in position 1: "
                       'invalid continuation byte',
                'type': 'value_error.unicodedecode'}]}
```

Trying to create a new user via the API and I am getting this error about the password, but it doesn't happen when using the Swagger docs so it must be something with the way I'm sending the request....

**Comment by ThomasLaPiana** [comment]
> running into a perplexing error:
> 
> ```
> fides> fides user create newusera Newpass!word1
> Loading config from: .fides/fides.toml
> {'username': 'newusera', 'password': 'Newpass!word1'}
> { 'detail': [ { 'loc': ['body', 'password'],
>                 'msg': "'utf-8' codec can't decode byte 0xec in position 1: "
>                        'invalid continuation byte',
>                 'type': 'value_error.unicodedecode'}]}
> ```
> 
> Trying to create a new user via the API and I am getting this error about the password, but it doesn't happen when using the Swagger docs so it must be something with the way I'm sending the request....

@sanders41 I confirmed that that password specifically was causing the issue, if I make the same call with `HolyCow1*1` it works perfectly.

Any idea why? What is weird about `Newpassword1!`?

**Comment by sanders41** [comment]
I’m on my phone instead of computer at the moment so I can’t test this, but I think the issue, may be that the endpoint expects to receive the password as a base64 encoded string (see [here](https://github.com/ethyca/fides/blob/903c06c8af575866e728633b647c43e0e54383a4/src/fides/lib/oauth/schemas/user.py#L61)). So I am thinking maybe it can decode `HolyCow1*1` as if it was base64 encoded but not `Newpassword1!`

**Comment by ThomasLaPiana** [comment]
> I’m on my phone instead of computer at the moment so I can’t test this, but I think the issue, may be that the endpoint expects to receive the password as a base64 encoded string (see [here](https://github.com/ethyca/fides/blob/903c06c8af575866e728633b647c43e0e54383a4/src/fides/lib/oauth/schemas/user.py#L61)). So I am thinking maybe it can decode `HolyCow1*1` as if it was base64 encoded but not `Newpassword1!`

Ah I see......that'd do it

File: tests/ctl/cli/test_cli.py [hunk 1]
```diff
@@ -169,678 +171,735 @@ def test_audit(test_config_path: str, test_cli_runner: CliRunner) -> None:
...
-            dataset_name,
-            f"{tmp_output_file}",
-            "--credentials-id",
-            "bigquery_1",
-        ],
-    )
-    print(result.output)
-    assert result.exit_code == 0
+    @pytest.mark.unit
+    def test_user_login_provide_credentials(
```
Comment (sanders41) [inline tests/ctl/cli/test_cli.py:856]: I think we should add a way to patch `CREDENTIALS_FILE_PATH` at test time and use a fixture to patch the path with `tmp_path` so we don't write test data to user's actual home directory.

**Review by sanders41 — CHANGES_REQUESTED** [review]

File: src/fides/core/user.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    response = requests.post(server_url + LOGIN_PATH, json=payload)
+    handle_cli_response(response, verbose=False)
+    user_id: str = response.json()["user_data"]["id"]
+    access_token: str = response.json()["token_data"]["access_token"]
+    return (user_id, access_token)
+
+
+def write_credentials_file(
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/core/user.py:49]: This is intended. This is a credentials file after all, and is purposefully stored in the user's home directory. This follows the pattern of other CLI tools that use authentication

File: tests/ctl/cli/test_cli.py (hunk 1 above)
Comment (ThomasLaPiana) [reply tests/ctl/cli/test_cli.py:856]: that's a `to do` up above :) working on it today

File: src/fides/cli/commands/user.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        server_url=server_url,
+    )
+    user_id = user_response.json()["id"]
+    update_user_permissions(
+        user_id=user_id, scopes=SCOPES, auth_header=auth_header, server_url=server_url
+    )
+    echo_green(f"User: '{username}' created and assigned permissions.")
+
+
+@user.command()
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/cli/commands/user.py:78]: this doesn't leverage click's built-in support for this which reduces code complexity a bit, will post my solution below

**Comment by ThomasLaPiana** [comment]
### New login flow:

No options provided
```
fides> fides user login
Loading config from: .fides/fides.toml
Username: root_user
Password: 
Logged in as user: root_user
Credentials file written to: C:\Users\tlapiana/.fides_credentials
```

Username provided only
```
fides> fides user login -u root_user
Loading config from: .fides/fides.toml
Password: 
Logged in as user: root_user
Credentials file written to: C:\Users\tlapiana/.fides_credentials

Password provided only
```
fides> fides user login -p Testpassword1!
Loading config from: .fides/fides.toml
Username: root_user
Logged in as user: root_user
Credentials file written to: C:\Users\tlapiana/.fides_credentials
```

File: src/fides/cli/commands/user.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanders41) [reply src/fides/cli/commands/user.py:46]: Perfect, solves the concern in a much cleaner way!

**Review by sanders41 — APPROVED** [review]

**Pull request merged**